ocdata/core/surface.py

```diff
@@ -0,0 +1,275 @@
+        """
+        This function fixes sub-surface atoms of a surface. Also works on systems
+        that have surface + adsorbate(s), as long as the bulk atoms are tagged with
+        `0`, surface atoms are tagged with `1`, and the adsorbate atoms are tagged
+        with `2` or above.
+
+        This function is used for both surface atoms and the combined surface+adsorbate
+
+        Inputs:
+            atoms           `ase.Atoms` class of the surface system. The tags of
+                            these atoms must be set such that any bulk/surface
+                            atoms are tagged with `0` or `1`, resectively, and any
+                            adsorbate atom is tagged with a 2 or above.
+        Returns:
+            atoms           A deep copy of the `atoms` argument, but where the appropriate
+                            atoms are constrained.
+        """
+        # We'll be making a `mask` list to feed to the `FixAtoms` class. This
+        # list should contain a `True` if we want an atom to be constrained, and
+        # `False` otherwise.
+        mask = [True if atom.tag == 0 else False for atom in self.atoms]
+        self.atoms.constraints += [FixAtoms(mask=mask)]
+
+    def __len__(self):
+        return len(self.atoms)
+
+    def __str__(self):
+        return f"Surface: {self.atoms.get_chemical_formula()}"
+
+    def __repr__(self):
+        return self.__str__()
+
+
+def tile_atoms(atoms, min_xy=MIN_XY):
+    """
+    This function will repeat an atoms structure in the x and y direction until
+    the x and y dimensions are at least as wide as the MIN_XY constant.
+
+    Arguments
+    ---------
+    atoms: ase.Atoms
+        The structure to tile.
+
+    Returns
+    -------
+    atoms_tiled: ase.Atoms
+        The tiled structure.
+    """
+    x_length = np.linalg.norm(atoms.cell[0])
+    y_length = np.linalg.norm(atoms.cell[1])
+    nx = int(math.ceil(min_xy / x_length))
+    ny = int(math.ceil(min_xy / y_length))
+    n_xyz = (nx, ny, 1)
+    atoms_tiled = atoms.repeat(n_xyz)
+    return atoms_tiled
+
+
+def find_surface_atoms_by_height(surface_atoms):
+    """
+    As discussed in the docstring for `find_surface_atoms_with_voronoi`,
+    sometimes we might accidentally tag a surface atom as a bulk atom if there
+    are multiple coordination environments for that atom type within the bulk.
+    One heuristic that we use to address this is to simply figure out if an
+    atom is close to the surface. This function will figure that out.
+
+    Specifically:  We consider an atom a surface atom if it is within 2
+    Angstroms of the heighest atom in the z-direction (or more accurately, the
+    direction of the 3rd unit cell vector).
+
+    Arguments
+    ---------
+    surface_atoms: ase.Atoms
+
+    Returns
+    -------
+    tags: list
+        A list that contains the indices of the surface atoms.
+    """
+    unit_cell_height = np.linalg.norm(surface_atoms.cell[2])
```

Comment:
Probably not a big deal but `cell[2]` is not necessarily in the (0,0,1) direction. so this is not the slab height in those cases.